_source/_posts/2020-12-08-beginners-guide-to-jwt.adoc

```diff
@@ -0,0 +1,175 @@
+
+TIP: JWT is pronounced "jot".
+
+toc::[]
+
+== What is a JWT
+
+A JWT is a structured security token format used to encode JSON data. Tokens can be cryptographically verified to ensure they have not been changed. There are two types of JWTs:
+
+- **J**SON **W**eb **S**ignature (JWS)
+- **J**SON **W**eb **E**ncryption (JWE)
+
+The data in a JWS is public, meaning anyone with the token can read the data, whereas a JWE is encrypted and private.  To read data contained within a JWE, you need both the token and a secret key.
+
+Most of the time, when you JWTs are used, it's actually JWS. The 'S' (the signature) is the important part and allows the token to be validated.  For the rest of this post, I'm going to talk about JWS and only lightly touch on the differences of a JWE.
+
+== How JWTs Are Used
+
+OAuth 2.0 identity providers (IdP) commonly use JWTs for access tokens.   You may have seen an HTTP request with an authorization header that looks like this:
+
+[source,http]
+----
+Authorization: Bearer eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0.5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U
+----
+
+NOTE: JWT access tokens are NOT part of the OAuth 2.0 specification, but almost all IdPs support them.
+
+Using a JWT (actually a JWS) allows the token to be validated locally, without making an HTTP request back to the IdP, thereby increasing your application's performance.  Applications can make use of data inside the token, further reducing expensive HTTP calls and database lookups.
+
+== JWT structure
+
+A JWS (the most common type of JWT) contains three parts separated by a dot (`.`). The first two parts (the "header" and "payload") are Base64-URL encoded JSON, and the third is a cryptographic signature.
+
+Let's look at an example JWT:
+
+[source,txt]
+----
+eyJhbGciOiJIUzI1NiJ9.eyJuYW1lIjoiSm9lIENvZGVyIn0.5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U
+----
+
+Breaking this down into the individual sections we have:
+
+[source,sh]
+----
+eyJhbGciOiJIUzI1NiJ9 # header
+.
+eyJuYW1lIjoiSm9lIENvZGVyIn0 # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: If you have a JWT with more than three sections, it's probably a JWE.
+
+Next, each of the first two sections are base64-url decoded:
+
+[source,json]
+----
+{"alg":"HS256"} # header
+.
+{"name":"Joe Coder"} # payload
+.
+5dlp7GmziL2QS06sZgK4mtaqv0_xX4oFUuTDh1zHK4U #signature
+----
+
+TIP: On macOS  (with "coreutils" installed, you can base64-url decode the strings on the command line with: `echo eyJhbGciOiJIUzI1NiJ9 | gbasenc -d --base64url`
+
+The last section in the JWT, the signature, is also base64-url encoded, but it's just binary data; if you try to decode it, you will end up with non-displayable characters:
+
+```txt
+��i�i�￹�￹KN�f￹�￹֪�O�_￹R��￹\�+￹
```

Comment:
If you have time, I'd show what this looks like with `hexdump` ... I'd also show how to re-create it yourself using sha256 and the HS256 shared secret